functorch/_src/aot_autograd.py

```diff
@@ -513,41 +1011,81 @@ def add_dupe_args(args):
+        # Corresponds to number of inputs that only have their metadata mutated
+        num_mutated_data_inputs = _num_mutated_data_inputs
+        # Corresponds to number of inputs that get their metadata (but not data) mutated
+        # We don't return these in the compiled fw, and instead we stash enough info
+        # to replay the metadata mutations later.
+        num_mutated_metadata_only_inputs = _num_mutated_metadata_only_inputs
+        # Corresponds to number of outputs in the original fw that are aliases of inputs
+        # (These are all not returned by the compiled forward, and instead they are manually
+        # created in the epilogue)
+        num_aliased_outputs = _num_aliased_outputs
+        # See Note [Outputs that alias mutated inputs]
+        # TODO: we don't need this! delete
+        num_output_aliases_from_mutated_inps = _num_output_aliases_from_mutated_inps
+        synthetic_base_info = _synthetic_base_info
+        fw_metadata = _fw_metadata
 
         @staticmethod
         @disable_torchdynamo
         def forward(ctx, *deduped_flat_tensor_args):
+
+            # There is a pretty complicated calling convention around what the compiled fw returns.
+            # The full list of outputs and their relative order is:
+            # (*mutated_data_inputs, *non_aliased_fw_outs, *saved_tensors, *saved_symints)
+            # - Note that in the synthetic bases case, mutated_inputs will correspond to an updated version
+            #   of the original view, and not the synthetic base
             fw_outs = call_func_with_args(
                 CompiledFunction.compiled_fw, deduped_flat_tensor_args, disable_amp=disable_amp
             )
-            num_outs = CompiledFunction.num_outs
+
+            num_non_aliased_outs = CompiledFunction.num_non_aliased_outs
             num_symints = CompiledFunction.num_symints
+            num_mutated_data_inputs = CompiledFunction.num_mutated_data_inputs
+            # Our forward() returns both outputs and mutated inputs,
+            num_forward_returns = num_mutated_data_inputs + num_non_aliased_outs
+
             # Partitioners must put symint arguments at the end separate from tensor arguments
             if num_symints > 0:
-                ctx.save_for_backward(*fw_outs[num_outs:-num_symints])
+                ctx.save_for_backward(*fw_outs[num_forward_returns:-num_symints])
                 ctx.symints = fw_outs[-num_symints:]
             else:
-                ctx.save_for_backward(*fw_outs[num_outs:])
+                ctx.save_for_backward(*fw_outs[num_forward_returns:])
                 ctx.symints = []
 
             fw_outs_not_requiring_grad = [
-                x for (i, x) in enumerate(fw_outs[0:num_outs]) if CompiledFunction.flat_outs_not_requiring_grad[i]
+                x for (i, x) in enumerate(fw_outs[:num_forward_returns])
+                if isinstance(x, torch.Tensor) and not CompiledFunction.fw_metadata.requires_grad_out_info[i]
```

Comment:
yeah... we could remove it if you think it's non-negligible perf.

We need it because the way that I set up the metadata was that `fw_metadata.requires_grad_info` is a giant `List[bool]`, where every output of the compiled forward graph gets an entry.

It's only true for outputs that do require gradient, and false for everything else (aka tensors that don't require grad, and also non-tensors).

If we want, I could change that list so that True represents "this output is a tensor that does not require gradients" and False is everything else.